## Business Understanding

### Problem Statements
Berdasarkan kondisi yang telah diuraikan sebelumnya, muncul dalam pertanyaan, seperti:
- Bagaimana sistem rekomendasi berbasis kesamaan kategori produk dapat membantu pengguna menemukan produk lain yang relevan?
- Bagaimana sistem ini dapat memberikan rekomendasi berdasarkan informasi kategori dan popularitas produk melalui rating?
- Bagaimana sistem ini dapat memberikan rekomendasi berdasarkan informasi kategori dan kualitas produk melalui total rekomendasi di tiap produk dari pengguna?

### Goals
Tujuan utama dari proyek analisis prediktif ini adalah untuk menjawab pertanyaan-pertanyaan di atas, beberapa tujuan spesifik yang ingin dicapai adalah sebagai berikut:
- Menyediakan rekomendasi produk yang relevan dengan teknik Content-Based Filtering menggunakan kesamaan kategori produk
- Memanfaatkan informasi kategori dan rating produk untuk menampilkan produk serupa yang lebih relevan bagi pengguna
- Memanfaatkan informasi kategori dan total rekomendasi tiap produk untuk menampilkan produk serupa yang lebih relevan bagi pengguna

### Solution statements
Dengan menggunakan Content-Based Filtering dengan TF-IDF dan Cosine Similarity

Pendekatan ini memanfaatkan representasi tekstual produk melalui teknik TF-IDF (Term Frequency-Inverse Document Frequency) untuk mengukur pentingnya kata-kata dalam kategori produk. Setiap produk akan direpresentasikan sebagai vektor berdasarkan kata kunci yang terdapat dalam kategori, dan kesamaan antar produk akan dihitung menggunakan cosine similarity. Dengan cara ini, sistem dapat merekomendasikan produk kepada pengguna berdasarkan kesamaan kategori dengan produk yang telah mereka pilih sebelumnya.

## Data Understanding

Dataset berisi 7.636 baris dan 19 kolom, dataset terdiri dari Kolom-kolomnya berisi brand_name, product_name, product_id, beauty_point_earned, price_range, price_by_combinations, url, active_date, default_category, categories, rating_types_str, average_rating, total_reviews, average_rating_by_types, total_recommended_count, total_repurchase_maybe_count, total_repurchase_no_count, total_repurchase_yes_count, total_in_wishlist. Variabel yang akan digunakan pada kasus kali ini sebagai parameter rekomendasi adalah variabel default_category. Kondisi data masih belum bersih dengan ditandai masih adanya missing values.

Referensi:
Hafizhan Ibrahim. "Sociolla: All Brands Products Catalog". Tautan: [https://www.kaggle.com/datasets/ibrahimhafizhan/sociolla-all-brands-products-catalog]. Diakses pada 27 Oktober 2024

## Data Loading

In [6]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
%matplotlib inline
import seaborn as sns

In [7]:
products = pd.read_csv('/content/products_all_brands.csv')
products

brand_name                  product_name  product_id  \
0             796_3ce       MULTI EYE COLOR PALETTE       97802   
1             796_3ce               VELVET LIP TINT       97810   
2             796_3ce                     LIP COLOR       97822   
3             796_3ce  MINI MULTI EYE COLOR PALETTE       97833   
4             796_3ce                    FACE BLUSH       97801   
...               ...                           ...         ...   
7631  273_yves-rocher    Ambre Noir Eau De Toilette        2158   
7632  273_yves-rocher  Soin Stimulating Conditioner       64174   
7633  273_yves-rocher              Repair Hair Mask       36357   
7634  273_yves-rocher   Hand Cream Olive Petitgrain       36404   
7635  273_yves-rocher                    Color Mask       64153   

      beauty_point_earned              price_range price_by_combinations  \
0                     130  Rp 555.000 - Rp 687.000                   NaN   
1                      50               Rp 264.000                   NaN   
2                      60               Rp 317.000                   NaN   
3                      80               Rp 423.000                   NaN   
4                      60               Rp 300.000                   NaN   
...                   ...                      ...                   ...   
7631                  130               Rp 659.000                   NaN   
7632                   30               Rp 199.000                   NaN   
7633                   50               Rp 289.000                   NaN   
7634                   20               Rp 109.000                   NaN   
7635                   50               Rp 289.000                   NaN   

                                                    url  \
0     https://www.sociolla.com/eyeshadow/69460-phan-...   
1     https://www.sociolla.com/lip-cream/69468-son-k...   
2     https://www.sociolla.com/lip-matte/69480-son-t...   
3     https://www.sociolla.com/eyeshadow/69491-phan-...   
4     https://www.sociolla.com/blush/69459-phan-ma-h...   
...                                                 ...   
7631  https://sociolla.com/fragrance/8088-ambre-noir...   
7632  https://vn.sociolla.com/d-u-x/17239-soin-stimu...   
7633  https://www.sociolla.com/hair-mask/14833-repai...   
7634  https://www.sociolla.com/221-hand-and-foot-cre...   
7635   https://vn.sociolla.com/kem-t-c/17224-color-mask   

                   active_date   default_category  \
0     2022-10-03T03:30:06.681Z          Eyeshadow   
1     2022-10-03T03:02:40.340Z          Lip Cream   
2     2023-05-30T09:49:15.158Z           Lipstick   
3     2022-10-03T03:27:01.334Z          Eyeshadow   
4     2022-10-03T03:22:26.610Z              Blush   
...                        ...                ...   
7631                       NaN    Eau De Toilette   
7632                       NaN        Conditioner   
7633  2022-10-27T07:33:19.380Z          Hair Mask   
7634                       NaN  Hand & Foot Cream   
7635  2023-09-20T10:38:15.154Z          Hair Mask   

                                            categories  \
0                              Makeup; Eyes; Eyeshadow   
1                              Makeup; Lips; Lip Cream   
2                               Makeup; Lips; Lipstick   
3                              Makeup; Eyes; Eyeshadow   
4                                  Makeup; Face; Blush   
...                                                ...   
7631                 Shop By Departments; Home; UOBMon   
7632                   Hair Care; Shampoo; Conditioner   
7633              Hair Care; Hair Treatment; Hair Mask   
7634  Shop By Departments; Home; 12.12 PRICE POINT 50%   
7635              Hair Care; Hair Treatment; Hair Mask   

                                       rating_types_str  average_rating  \
0     is_star_long_wear;is_star_packaging;is_star_pi...        4.920000   
1     is_star_long_wear;is_star_packaging;is_star_pi...        4.576190   
2     is_star_long_wear;i

- Data berhasil terpanggil, di sini data terdiri dari 7636 baris dan 19 kolom.
- Kolom-kolomnya berisi brand_name, product_name, product_id, beauty_point_earned, price_range, price_by_combinations, url, active_date, default_category, categories, rating_types_str, average_rating, total_reviews, average_rating_by_types, total_recommended_count, total_repurchase_maybe_count, total_repurchase_no_count, total_repurchase_yes_count, total_in_wishlist

## Univariate Exploratory Data Analysis

Variabel-variabel pada Sociolla: All Brands Products Catalog  dataset adalah sebagai berikut:
- brand_name : id dan merek atau nama brand dari tiap produk, yang dipisahkan dengan garis bawah
- product_name : nama produk
- product_id : id produk
- beauty_point_earned : poin kecantikan yang diperoleh melalui pembelian
- price_range : kisaran umum harga produk
- price_by_combinations : kisaran khusus harga produk berdasarkan variasi produk yang berbeda
- url : URL link yang mengarahkan pada laman Sociolla.com
- active_date : Informasi tentang tanggal setiap produk menjadi aktif atau tersedia di Sociolla.com
- default_category : Kategori umum produk
- categories : Kategori khusus produk, untuk klasifikasi produk terperinci
- rating_types_str : Uasan konsumen tiap produk
- average_rating : Rata-rata rating penilaian produk
- total_reviews : Total konsumen yang memberikan ulasan
- average_rating_by_types : Rata-rata rating penilaian produk dalam aspek tertentu
- total_recommended_count : Total konsumen yang merekomendasikan produk
- total_repurchase_maybe_count : Total konsumen yang mungkin membeli produk ulang
- total_repurchase_no_count : Total konsumen yang tidak membeli produk ulang
- total_repurchase_yes_count : Total konsumen yang membeli produk ulang
- total_in_wishlist : Total konsumen yang memasukkan produk ke dalam wishlist

Variabel default_category, average_rating, dan total_recommended_count akan digunakan pada model rekomendasi. Sedangkan, variabel brand_name, product_name, dan price_range untuk melihat output yang dihasilkan.


In [8]:
products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7636 entries, 0 to 7635
Data columns (total 19 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   brand_name                    7636 non-null   object 
 1   product_name                  7636 non-null   object 
 2   product_id                    7636 non-null   int64  
 3   beauty_point_earned           7636 non-null   int64  
 4   price_range                   7636 non-null   object 
 5   price_by_combinations         4087 non-null   object 
 6   url                           7636 non-null   object 
 7   active_date                   5534 non-null   object 
 8   default_category              7636 non-null   object 
 9   categories                    7632 non-null   object 
 10  rating_types_str              7561 non-null   object 
 11  average_rating                7636 non-null   float64
 12  total_reviews                 7636 non-null   int64  
 13  ave

In [9]:
print('Jumlah product_id: ', len(products.product_id.unique()))
print('Jumlah brand_name: ', len(products.brand_name.unique()))
print('Jumlah data average_rating: ', len(products.average_rating.unique()))
print('Jumlah data default_category: ', len(products.default_category.unique()))

Jumlah product_id:  7636
Jumlah brand_name:  321
Jumlah data average_rating:  3187
Jumlah data default_category:  195


Dengan fungsi unnique, dapat diketahui jika dataaset terdiri dari 7636 nama produk yang berbeda, 319 nama brand yang berbeda, 3187 nilai rating yang berbeda, dan 195 kategori produk yang berbeda

## Data Preparation

#### Memeriksa Data Terduplikasi

In [10]:
products.duplicated().sum()

0

Data tidak memiliki data terduplikasi

#### Mengatasi Missing Value

In [11]:
products.describe()

product_id  beauty_point_earned  average_rating  total_reviews  \
count    7636.000000          7636.000000     7636.000000    7636.000000   
mean    82903.974725            37.236773        3.517905     198.197617   
std     27257.173650            46.056962        2.000138     852.067365   
min        68.000000             0.000000        0.000000       0.000000   
25%     75267.000000            10.000000        3.910714       1.000000   
50%     91856.500000            20.000000        4.576923       9.000000   
75%    101825.500000            50.000000        4.738155      76.000000   
max    109155.000000           610.000000        5.000000   21536.000000   

       total_recommended_count  total_repurchase_maybe_count  \
count              7636.000000                   7636.000000   
mean                186.560503                     40.603457   
std                 803.185318                    169.473278   
min                   0.000000                      0.000000   
25%                   1.000000                      0.000000   
50%                   9.000000                      2.000000   
75%                  71.000000                     17.000000   
max               20804.000000                   5025.000000   

       total_repurchase_no_count  total_repurchase_yes_count  \
count                7636.000000                 7636.000000   
mean                   15.479963                  141.873494   
std                    79.716058                  625.673542   
min                     0.000000                    0.000000   
25%                     0.000000                    0.000000   
50%                     0.000000                    6.000000   
75%                     6.000000                   52.000000   
max                  2824.000000                17653.000000   

       total_in_wishlist  
count        7636.000000  
mean          633.016108  
std          2055.067002  
min          -317.000000  
25%            28.000000  
50%           132.000000  
75%           480.000000  
max         58568.000000

In [12]:
products.isnull().sum()

brand_name                         0
product_name                       0
product_id                         0
beauty_point_earned                0
price_range                        0
price_by_combinations           3549
url                                0
active_date                     2102
default_category                   0
categories                         4
rating_types_str                  75
average_rating                     0
total_reviews                      0
average_rating_by_types         1846
total_recommended_count            0
total_repurchase_maybe_count       0
total_repurchase_no_count          0
total_repurchase_yes_count         0
total_in_wishlist                  0
dtype: int64

terdapat banyak missing values di beberapa kolom, seperti 'price_by_combinations', 'active_date', dan 'average_rating_by_types'. Untuk itu, kolom-kolom ini akan di drop, karena memiliki missing values dan nilai yang tidak terlalu berpengaruh terhadap tujuan rekomendasi.


Kolom-kolom lainnya yang tidak berpengaruh juga akan di drop, seperti 'beauty_point_earned', 'active_date', 'categories', 'rating_types_str', 'total_repurchase_maybe_count', 'total_repurchase_no_count', 'total_repurchase_yes_count'

In [13]:
#Kolom yang tidak terlalu berpengaruh dan memiliki NaN didrop
products.drop(['beauty_point_earned'], inplace=True, axis=1)
products.drop(['price_by_combinations'], inplace=True, axis=1)
products.drop(['active_date'], inplace=True, axis=1)
products.drop(['categories'], inplace=True, axis=1)
products.drop(['rating_types_str'], inplace=True, axis=1)
products.drop(['average_rating_by_types'], inplace=True, axis=1)
products.drop(['total_repurchase_maybe_count'], inplace=True, axis=1)
products.drop(['total_repurchase_no_count'], inplace=True, axis=1)
products.drop(['total_repurchase_yes_count'], inplace=True, axis=1)
products

brand_name                  product_name  product_id  \
0             796_3ce       MULTI EYE COLOR PALETTE       97802   
1             796_3ce               VELVET LIP TINT       97810   
2             796_3ce                     LIP COLOR       97822   
3             796_3ce  MINI MULTI EYE COLOR PALETTE       97833   
4             796_3ce                    FACE BLUSH       97801   
...               ...                           ...         ...   
7631  273_yves-rocher    Ambre Noir Eau De Toilette        2158   
7632  273_yves-rocher  Soin Stimulating Conditioner       64174   
7633  273_yves-rocher              Repair Hair Mask       36357   
7634  273_yves-rocher   Hand Cream Olive Petitgrain       36404   
7635  273_yves-rocher                    Color Mask       64153   

                  price_range  \
0     Rp 555.000 - Rp 687.000   
1                  Rp 264.000   
2                  Rp 317.000   
3                  Rp 423.000   
4                  Rp 300.000   
...                       ...   
7631               Rp 659.000   
7632               Rp 199.000   
7633               Rp 289.000   
7634               Rp 109.000   
7635               Rp 289.000   

                                                    url   default_category  \
0     https://www.sociolla.com/eyeshadow/69460-phan-...          Eyeshadow   
1     https://www.sociolla.com/lip-cream/69468-son-k...          Lip Cream   
2     https://www.sociolla.com/lip-matte/69480-son-t...           Lipstick   
3     https://www.sociolla.com/eyeshadow/69491-phan-...          Eyeshadow   
4     https://www.sociolla.com/blush/69459-phan-ma-h...              Blush   
...                                                 ...                ...   
7631  https://sociolla.com/fragrance/8088-ambre-noir...    Eau De Toilette   
7632  https://vn.sociolla.com/d-u-x/17239-soin-stimu...        Conditioner   
7633  https://www.sociolla.com/hair-mask/14833-repai...          Hair Mask   
7634  https://www.sociolla.com/221-hand-and-foot-cre...  Hand & Foot Cream   
7635   https://vn.sociolla.com/kem-t-c/17224-color-mask          Hair Mask   

      average_rating  total_reviews  total_recommended_count  \
0           4.920000              5                        5   
1           4.576190             42                       42   
2           0.000000              0                        0   
3           4.883333              6                       12   
4           4.858824             13                       17   
...              ...            ...                      ...   
7631        4.476190              7                        6   
7632        4.638889              9                        7   
7633        4.750000              5                        5   
7634        4.968750              8                        8   
7635        4.500000              2                        2   

      total_in_wishlist  
0                   717  
1                   682  
2                   173  
3                   257  
4                   387  
...                 ...  
7631                355  
7632                122  
7633                 58  
7634                117  
7635                 35  

[7636 rows x 10 columns]

Dapat, dilihat, kolom-kolom yang tidak berpengaruh berhasil dihapus

In [14]:
products.isnull().sum()

brand_name                 0
product_name               0
product_id                 0
price_range                0
url                        0
default_category           0
average_rating             0
total_reviews              0
total_recommended_count    0
total_in_wishlist          0
dtype: int64

Setelah diperiksa, nilai missing values sudah tidak ada

#### Memisahkan Nilai Kolom Name Brand

Dapat diketahui jika, kolom 'brand_name', tergabung dari dua values, yaitu nama brand, dan id brand, untuk itu dilakukan pemisahan dengan fungsi split()

In [15]:
products[['brand_name_id', 'brand_name']] = products['brand_name'].str.split('_', expand=True)
products

brand_name                  product_name  product_id  \
0             3ce       MULTI EYE COLOR PALETTE       97802   
1             3ce               VELVET LIP TINT       97810   
2             3ce                     LIP COLOR       97822   
3             3ce  MINI MULTI EYE COLOR PALETTE       97833   
4             3ce                    FACE BLUSH       97801   
...           ...                           ...         ...   
7631  yves-rocher    Ambre Noir Eau De Toilette        2158   
7632  yves-rocher  Soin Stimulating Conditioner       64174   
7633  yves-rocher              Repair Hair Mask       36357   
7634  yves-rocher   Hand Cream Olive Petitgrain       36404   
7635  yves-rocher                    Color Mask       64153   

                  price_range  \
0     Rp 555.000 - Rp 687.000   
1                  Rp 264.000   
2                  Rp 317.000   
3                  Rp 423.000   
4                  Rp 300.000   
...                       ...   
7631               Rp 659.000   
7632               Rp 199.000   
7633               Rp 289.000   
7634               Rp 109.000   
7635               Rp 289.000   

                                                    url   default_category  \
0     https://www.sociolla.com/eyeshadow/69460-phan-...          Eyeshadow   
1     https://www.sociolla.com/lip-cream/69468-son-k...          Lip Cream   
2     https://www.sociolla.com/lip-matte/69480-son-t...           Lipstick   
3     https://www.sociolla.com/eyeshadow/69491-phan-...          Eyeshadow   
4     https://www.sociolla.com/blush/69459-phan-ma-h...              Blush   
...                                                 ...                ...   
7631  https://sociolla.com/fragrance/8088-ambre-noir...    Eau De Toilette   
7632  https://vn.sociolla.com/d-u-x/17239-soin-stimu...        Conditioner   
7633  https://www.sociolla.com/hair-mask/14833-repai...          Hair Mask   
7634  https://www.sociolla.com/221-hand-and-foot-cre...  Hand & Foot Cream   
7635   https://vn.sociolla.com/kem-t-c/17224-color-mask          Hair Mask   

      average_rating  total_reviews  total_recommended_count  \
0           4.920000              5                        5   
1           4.576190             42                       42   
2           0.000000              0                        0   
3           4.883333              6                       12   
4           4.858824             13                       17   
...              ...            ...                      ...   
7631        4.476190              7                        6   
7632        4.638889              9                        7   
7633        4.750000              5                        5   
7634        4.968750              8                        8   
7635        4.500000              2                        2   

      total_in_wishlist brand_name_id  
0                   717           796  
1                   682           796  
2                   173           796  
3                   257           796  
4                   387           796  
...                 ...           ...  
7631                355           273  
7632                122           273  
7633                 58           273  
7634                117           273  
7635                 35           273  

[7636 rows x 11 columns]

Hasilnya pun, nilai berhasil dipisah, dan kolom 'brand_name_id' berhasil dibuat

#### TF-IDF Vectorizer

Teknik TF-IDF digunakan pada sistem rekomendasi untuk menemukan representasi fitur penting dari setiap kategori produk.

In [16]:
products

brand_name                  product_name  product_id  \
0             3ce       MULTI EYE COLOR PALETTE       97802   
1             3ce               VELVET LIP TINT       97810   
2             3ce                     LIP COLOR       97822   
3             3ce  MINI MULTI EYE COLOR PALETTE       97833   
4             3ce                    FACE BLUSH       97801   
...           ...                           ...         ...   
7631  yves-rocher    Ambre Noir Eau De Toilette        2158   
7632  yves-rocher  Soin Stimulating Conditioner       64174   
7633  yves-rocher              Repair Hair Mask       36357   
7634  yves-rocher   Hand Cream Olive Petitgrain       36404   
7635  yves-rocher                    Color Mask       64153   

                  price_range  \
0     Rp 555.000 - Rp 687.000   
1                  Rp 264.000   
2                  Rp 317.000   
3                  Rp 423.000   
4                  Rp 300.000   
...                       ...   
7631               Rp 659.000   
7632               Rp 199.000   
7633               Rp 289.000   
7634               Rp 109.000   
7635               Rp 289.000   

                                                    url   default_category  \
0     https://www.sociolla.com/eyeshadow/69460-phan-...          Eyeshadow   
1     https://www.sociolla.com/lip-cream/69468-son-k...          Lip Cream   
2     https://www.sociolla.com/lip-matte/69480-son-t...           Lipstick   
3     https://www.sociolla.com/eyeshadow/69491-phan-...          Eyeshadow   
4     https://www.sociolla.com/blush/69459-phan-ma-h...              Blush   
...                                                 ...                ...   
7631  https://sociolla.com/fragrance/8088-ambre-noir...    Eau De Toilette   
7632  https://vn.sociolla.com/d-u-x/17239-soin-stimu...        Conditioner   
7633  https://www.sociolla.com/hair-mask/14833-repai...          Hair Mask   
7634  https://www.sociolla.com/221-hand-and-foot-cre...  Hand & Foot Cream   
7635   https://vn.sociolla.com/kem-t-c/17224-color-mask          Hair Mask   

      average_rating  total_reviews  total_recommended_count  \
0           4.920000              5                        5   
1           4.576190             42                       42   
2           0.000000              0                        0   
3           4.883333              6                       12   
4           4.858824             13                       17   
...              ...            ...                      ...   
7631        4.476190              7                        6   
7632        4.638889              9                        7   
7633        4.750000              5                        5   
7634        4.968750              8                        8   
7635        4.500000              2                        2   

      total_in_wishlist brand_name_id  
0                   717           796  
1                   682           796  
2                   173           796  
3                   257           796  
4                   387           796  
...                 ...           ...  
7631                355           273  
7632                122           273  
7633                 58           273  
7634                117           273  
7635                 35           273  

[7636 rows x 11 columns]

In [17]:
from sklearn.feature_extraction.text import TfidfVectorizer

tf = TfidfVectorizer()
tf.fit(products['default_category'])
tf.get_feature_names_out()

array(['2in1', 'accessories', 'acne', 'after', 'ampoule', 'and',
       'applicators', 'aromatherapy', 'arts', 'baby', 'bags', 'balm',
       'bar', 'bath', 'bb', 'beauty', 'blotting', 'blush', 'body',
       'booster', 'bottles', 'breast', 'bronzer', 'brow', 'brush',
       'brushes', 'bug', 'bundles', 'butter', 'cake', 'cap', 'card',
       'care', 'case', 'cc', 'cellulite', 'clay', 'cleaner', 'cleanser',
       'cleansing', 'clipper', 'clippers', 'clothing', 'color', 'combs',
       'concealer', 'conditioner', 'contour', 'cotton', 'crayon', 'cream',
       'curler', 'curling', 'cushion', 'de', 'deodorant', 'diaper', 'dry',
       'dryers', 'eau', 'essence', 'exclusive', 'exfoliants',
       'exfoliator', 'eye', 'eyebrows', 'eyelash', 'eyelashes',
       'eyeliner', 'eyeshadow', 'fabric', 'face', 'facial', 'false',
       'family', 'feminine', 'files', 'foam', 'foot', 'for', 'foundation',
       'fragrance', 'gel', 'gift', 'gloss', 'glue', 'hair', 'hand',
       'head', 'highlighter'

In [18]:
tfidf_matrix = tf.fit_transform(products['default_category'])
tfidf_matrix.shape

(7636, 210)

Matriks berukuran (7636, 210). Nilai 7636 merupakan ukuran data dan 210 merupakan matrik kategori produk atau banyaknya tipe dari kategori produk

In [19]:
tfidf_matrix.todense()

matrix([[0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        ...,
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.]])

Selanjutnya, matriks tf-idf untuk nama produk dan kategori produk.

In [20]:
pd.DataFrame(
    tfidf_matrix.todense(),
    columns=tf.get_feature_names_out(),
    index=products['product_name'].values
).sample(20, axis=1).sample(20, axis=0)

exfoliator  brow   shaving  \
Gillette Foamy Shave Cream Menthol                  0.0   0.0  0.808244   
8P Sponge Wedges Nude SBR                           0.0   0.0  0.000000   
New Softbrow Pensil                                 0.0   0.0  0.000000   
Duo Hypergloss Free Keychain + Mini SBWC            0.0   0.0  0.000000   
Sleek eyebrow set                                   0.0   0.0  0.000000   
Royal Lip Mousse                                    0.0   0.0  0.000000   
Apply & Reapply Sunscreen as Easy as 1-2-3!         0.0   0.0  0.000000   
AlpHA Deodorant Brightening Underarm Care           0.0   0.0  0.000000   
Glozee - Angelic Grey                               0.0   0.0  0.000000   
Body Wash PATCHOULI                                 0.0   0.0  0.000000   
Cenil Ubi Ungu Spot On Manicure                     0.0   0.0  0.000000   
Moist & Clean Skin                                  0.0   0.0  0.000000   
SRM NGUA MUN CX TRAI MO                             0.0   0.0  0.000000   
Rose Bloom Petal Essence                            0.0   0.0  0.000000   
Texture Experience Shampoo Green Tea Butter         0.0   0.0  0.000000   
B1G1 Midnight Sky - BHA Toner                       0.0   0.0  0.000000   
Artclass by Rodin Blusher                           0.0   0.0  0.000000   
Dr. Mascara Fixer                                   0.0   0.0  0.000000   
Atmosphere CC Cream                                 0.0   0.0  0.000000   
Bug Off Organic Mosquito Repellent Spray            0.0   0.0  0.000000   

                                             matte  paper  treatment  tools  \
Gillette Foamy Shave Cream Menthol             0.0    0.0        0.0    0.0   
8P Sponge Wedges Nude SBR                      0.0    0.0        0.0    0.0   
New Softbrow Pensil                            0.0    0.0        0.0    0.0   
Duo Hypergloss Free Keychain + Mini SBWC       0.0    0.0        0.0    0.0   
Sleek eyebrow set                              0.0    0.0        0.0    0.0   
Royal Lip Mousse                               0.0    0.0        0.0    0.0   
Apply & Reapply Sunscreen as Easy as 1-2-3!    0.0    0.0        0.0    0.0   
AlpHA Deodorant Brightening Underarm Care      0.0    0.0        0.0    0.0   
Glozee - Angelic Grey                          0.0    0.0        0.0    0.0   
Body Wash PATCHOULI                            0.0    0.0        0.0    0.0   
Cenil Ubi Ungu Spot On Manicure                0.0    0.0        0.0    0.0   
Moist & Clean Skin                             0.0    0.0        0.0    0.0   
SRM NGUA MUN CX TRAI MO                        0.0    0.0        0.0    0.0   
Rose Bloom Petal Essence                       0.0    0.0        0.0    0.0   
Texture Experience Shampoo Green Tea Butter    0.0    0.0        0.0    0.0   
B1G1 Midnight Sky - BHA Toner                  0.0    0.0        0.0    0.0   
Artclass by Rodin Blusher                      0.0    0.0        0.0    0.0   
Dr. Mascara Fixer                              0.0    0.0        0.0    0.0   
Atmosphere CC Cream                            0.0    0.0        0.0    0.0   
Bug Off Organic Mosquito Repellent Spray       0.0    0.0        0.0    0.0   

                                             toe  lotion      face  eyeliner  \
Gillette Foamy Shave Cream Menthol           0.0     0.0  0.000000       0.0   
8P Sponge Wedges Nude SBR                    0.0     0.0  0.000000       0.0   
New Softbrow Pensil                          0.0     0.0  0.000000       0.0   
Duo Hypergloss Free Keychain + Mini SBWC     0.0     0.0  0.000000       0.0   
Sleek eyebrow set                            0.0     0.0  0.000000       0.0   
Royal Lip Mousse                             0.0     0.0  0.000000       0.0   
Apply & Reapply Sunscreen as Easy as 1-2-3!  0.0     0.0  0.000000       0.0   
AlpHA Deodorant Brightening Underarm Care    0.0     0.0  0.000000       0.0   
Glozee - Angelic Grey                        0.0     0.0  0.000000       0.0   
Body Wash

Output sampel matriks tf-idf di atas menunjukkan Restaurant Shell Glow Highlighter (Like Shell Collection) termasuk dalam kategori highlighter. Hal ini terlihat dari nilai matriks 1.0 pada kategori highlighter. Begitu juga dengan Aloe Hydramild Facial Wash, termasuk dalam kategori face, dengan nilai 0.589674.

## Model Development dengan Content Based Filtering berdasarkan Kategori Produk

Teknik content based filtering akan merekomendasikan item yang mirip dengan item yang disukai pengguna di masa lalu. Pada tahap ini, ditemukan representasi fitur penting dari setiap kategori produk dengan tfidf vectorizer dan menghitung tingkat kesamaan dengan cosine similarity.

#### Cosine Similarity

Sekarang, dihitung derajat kesamaan (similarity degree) antar nama produk dengan teknik cosine similarity.

In [21]:
from sklearn.metrics.pairwise import cosine_similarity
cosine_sim = cosine_similarity(tfidf_matrix)
cosine_sim

array([[1.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 1.        , 0.        , ..., 0.        , 0.24144692,
        0.        ],
       [0.        , 0.        , 1.        , ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.        , 0.        , 0.        , ..., 1.        , 0.        ,
        1.        ],
       [0.        , 0.24144692, 0.        , ..., 0.        , 1.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 1.        , 0.        ,
        1.        ]])

Dihitung cosine similarity dataframe tfidf_matrix yang  diperoleh pada tahapan sebelumnya. Dengan satu baris kode untuk memanggil fungsi cosine similarity dari library sklearn, hasil similarity tiap nama produk sudah didapat, hasilnya berupa matriks kesamaan dalam bentuk array.

In [22]:
cosine_sim_df = pd.DataFrame(cosine_sim, index=products['product_name'], columns=products['product_name'])
print('Shape:', cosine_sim_df.shape)

cosine_sim_df.sample(20, axis=1).sample(20, axis=0)

Shape: (7636, 7636)


product_name                                        Your Skin Bae Serum Lactic Acid 10% + Niacinamide 2,5%  \
product_name                                                                                                 
Sekkisei Clear Treatment Essence                                                             0.000000        
Set Of 5 Soothing Serum Facial Mask                                                          0.000000        
Pond's Men Acne Solution Facial Foam                                                         0.364334        
Texture Experience Shampoo + Conditioner Mint S...                                           0.000000        
Dancoly texture paste (medium hold) - for men                                                0.000000        
2 Step brightening                                                                           0.000000        
MC2 Advance Soothing Essence Toner                                                           0.000000        
Joylab Package - Wonderskin Series                                                           0.000000        
Bonavie Package Poison of Maison du Safran                                                   0.000000        
Texture Experience Creambath Strawberry Yoghurt...                                           0.000000        
Pretty Relax Set                                                                             0.000000        
Ginger Essence Hair & Scalp Tonic                                                            0.000000        
Kundal Protein Treatment                                                                     0.000000        
Penelope  - Classic Collection                                                               0.000000        
Premium Moist & Repair Shampoo                                                               0.000000        
Clay Mask Duo                                                                                0.000000        
Hygiene Kit YLANG 3x60ml FREE Pouch Canvas                                                   0.000000        
Daily Moisture Lotion Twin Set                                                               0.313938        
Cleanse & Refresh Cucumber Facial Wipes                                                      0.000000        
Neroli Blossom Facial C-Serum                                                                1.000000        

product_name                                        Extra Bright Vitamin Lotion C&E 180mL Twinpack  \
product_name                                                                                         
Sekkisei Clear Treatment Essence                                                          0.000000   
Set Of 5 Soothing Serum Facial Mask                                                       0.000000   
Pond's Men Acne Solution Facial Foam                                                      0.278937   
Texture Experience Shampoo + Conditioner Mint S...                                        0.000000   
Dancoly texture paste (medium hold) - for men                                             0.000000   
2 Step brightening                                                                        0.000000   
MC2 Advance Soothing Essence Toner                                                        0.000000   
Joylab Package - Wonderskin Series                                                        0.000000   
Bonavie Package Poison of Maison du Safran                                                0.000000   
Texture Experience Creambath Strawberry Yoghurt...                                        0.000000   
Pretty Relax Set                                                                          0.000000   
Ginger Essence Hair & Scalp Tonic                                                         0.000000   
Kundal Protein Treatment                                                                  0.000000   
Penelope  - Classic Collection                   

Dengan cosine similarity, berhasil mengidentifikasi kesamaan antar produk lainnya. Shape (7636, 7636) merupakan ukuran matriks similarity dari data. Namun, dalam output, hanya menampilkan sebanyak 20 sampel data saja.


Contoh: angka 1.0 pada Skin Buddy Dot Burst Face Wash dan Bio Renew Deep Cleanser menunjukkan dua produk ini memiliki kesamaan. Begitu juga, dengan Oh! So Bright Serum dengan Skin'o'tic Serum, yang juga mendapat nilai 1.0

#### Mendapatkan Rekomendasi

Sistem rekomendasi akan memberikan produk yang memiliki similarity terhadap produk yang diinput oleh pengguna berdasarkan kesamaan kategori dari produk-produk rekomendasi, hasil similarity tiap produk sudah didapat dari perhitungan sebelumnya.

In [23]:
def products_recommendations(nama_produk, similarity_data=cosine_sim_df, items=products[['product_name',
                                                                                          'brand_name',
                                                                                          'price_range',]], k=10):
    index = similarity_data.loc[:,nama_produk].to_numpy().argpartition(
        range(-1, -k-1, -1))

    closest = similarity_data.columns[index[-1:-(k+2):-1]]
    closest = closest.drop(nama_produk, errors='ignore')
    recommendations = pd.DataFrame(closest, columns=['product_name']).merge(items, on='product_name')

    return recommendations.head(k)

Dibuat fungsi dengan nama products_recommendations, dengan nama_produk sebagai parameter pencarian, hasil kesamaan yang diambil dari cosine_sim_df, isi dari dataframe yang ingin ditampilkan dan k (jumlah rekomendasi yang diinginkan) sebanyak 10.

Lalu membuat index untuk mengambil urutan indek produk yang paling mirip, dengan similarity_data, dengan fungsi argpartition untuk mengurutkan indeks array berdasarkan skor kemiripan dari yang tertinggi ke terendah (berdasarkan parameter range(-1, -k-1, -1)).

Daftar produk disimpan dalam closest, dengan mengambil skor kemiripan tertinggi, lalu menghapus nama produk itu sendiri dari daftar rekomendasi. terakhir, dibuat variabel recommendation, untuk membuat dataframe dari data closest, column, dan items untuk digabung menjadi satu, dan nilai k dikembalikan.

In [24]:
products['product_name'].unique()

array(['MULTI EYE COLOR PALETTE', 'VELVET LIP TINT', 'LIP COLOR', ...,
       'Repair Hair Mask', 'Hand Cream Olive Petitgrain', 'Color Mask'],
      dtype=object)

In [25]:
products[products.product_name.eq('Repair Hair Mask')]

brand_name      product_name  product_id price_range  \
7633  yves-rocher  Repair Hair Mask       36357  Rp 289.000   

                                                    url default_category  \
7633  https://www.sociolla.com/hair-mask/14833-repai...        Hair Mask   

      average_rating  total_reviews  total_recommended_count  \
7633            4.75              5                        5   

      total_in_wishlist brand_name_id  
7633                 58           273

Output rekomendasi diharapkan akan memberikan produk serupa Repair Hair Mask, dengan kategori Hair Mask yang mirip

In [26]:
products_recommendations('Repair Hair Mask')

product_name             brand_name  \
0                                         Color Mask            yves-rocher   
1                    Extra Fresh & Hydrate Treatment            moist-diane   
2                            Moist & Shine Treatment            moist-diane   
3                      Bonheur Grasse Rose Treatment            moist-diane   
4                      Extra Damage Repair Hair Mask            moist-diane   
5                  Extra Smooth & Straight Treatment            moist-diane   
6                  Extra Smooth & Straight Hair Mask            moist-diane   
7  Texture Experience Creambath Vanilla Milk Inte...  makarizo-professional   
8                     Extra Volume & Scalp Treatment            moist-diane   
9  Bonheur Blue Jasmine Treatment Damage Repair &...            moist-diane   

  price_range  
0  Rp 289.000  
1  Rp 136.000  
2  Rp 136.000  
3  Rp 190.000  
4  Rp 119.000  
5  Rp 136.000  
6  Rp 119.000  
7  Rp 288.600  
8  Rp 136.000  
9  Rp 190.000

Output pun menampilkan 10 daftar produk rekomendasi yang memiliki kemiripan dengan Repair Hair Mask, dengan rentang harga dan asal brandnya juga

Menguji dengan nama produk lain

In [27]:
products_recommendations('LIP COLOR')

product_name        brand_name price_range
0            Lip Velvet Hydrating Balm      raine-beauty  Rp 154.000
1        Color Hypnose Creamy Lipmatte         make-over   Rp 89.000
2            PLUMP HIGH SHINE LIP GLOW         focallure   Rp 32.000
3                  Sugar Rush Lipstick             emina   Rp 39.500
4              Better Lips-Talk Velvet             etude  Rp 188.000
5                  PURE MATTE LIPSTICK         focallure   Rp 42.000
6                           Lip Bullet        blp-beauty  Rp 129.000
7       The Grail Vivid Matte Lipstick  mineral-botanica   Rp 43.900
8       Everyday Velvet Rouge Lipstick     buttonscarves  Rp 295.000
9  Stick With Me Velvet Matte Lipstick       nama-beauty  Rp 119.000

In [28]:
products_recommendations('Centella Mask Pack')

product_name brand_name price_range
0                 Missha Double Green Tea Pack   sociolla  Rp 148.800
1               MEDIHEAL Teatree Nude Gel Mask   mediheal   Rp 49.900
2                       Teatree Essential Mask   mediheal   Rp 29.900
3              MEDIHEAL Collagen Nude Gel Mask   mediheal   Rp 49.900
4                      Lemonlime Vita Ade Mask   mediheal   Rp 21.500
5              Gentle Natural Booster Set of 5   mediheal  Rp 149.500
6                       THE N.M.F Ampoule Mask   mediheal   Rp 29.900
7  BUY 2 GET 1 MEDIHEAL TEATREE ESSENTIAL MASK   mediheal   Rp 89.700
8  BUY 2 GET 1 MEDIHEAL THE N.M.F AMPOULE MASK   mediheal   Rp 89.700
9                      Collagen Essential Mask   mediheal   Rp 29.900

## Model Development dengan Content Based Filtering berdasarkan Rating Produk

Teknik content based filtering akan merekomendasikan item yang mirip dengan item yang disukai pengguna di masa lalu. Pada tahap ini, ditemukan representasi fitur penting dari setiap kategori produk dengan tfidf vectorizer dan menghitung tingkat kesamaan berdasarkan rating dengan cosine similarity.

### Data Preparation

##### TF-IDF Vectorizer

Teknik TF-IDF digunakan pada sistem rekomendasi untuk menemukan representasi fitur penting dari setiap kategori produk.

In [29]:
from sklearn.feature_extraction.text import TfidfVectorizer

tf = TfidfVectorizer()
tf.fit(products['default_category'])
tf.get_feature_names_out()

array(['2in1', 'accessories', 'acne', 'after', 'ampoule', 'and',
       'applicators', 'aromatherapy', 'arts', 'baby', 'bags', 'balm',
       'bar', 'bath', 'bb', 'beauty', 'blotting', 'blush', 'body',
       'booster', 'bottles', 'breast', 'bronzer', 'brow', 'brush',
       'brushes', 'bug', 'bundles', 'butter', 'cake', 'cap', 'card',
       'care', 'case', 'cc', 'cellulite', 'clay', 'cleaner', 'cleanser',
       'cleansing', 'clipper', 'clippers', 'clothing', 'color', 'combs',
       'concealer', 'conditioner', 'contour', 'cotton', 'crayon', 'cream',
       'curler', 'curling', 'cushion', 'de', 'deodorant', 'diaper', 'dry',
       'dryers', 'eau', 'essence', 'exclusive', 'exfoliants',
       'exfoliator', 'eye', 'eyebrows', 'eyelash', 'eyelashes',
       'eyeliner', 'eyeshadow', 'fabric', 'face', 'facial', 'false',
       'family', 'feminine', 'files', 'foam', 'foot', 'for', 'foundation',
       'fragrance', 'gel', 'gift', 'gloss', 'glue', 'hair', 'hand',
       'head', 'highlighter'

In [30]:
tfidf_matrix = tf.fit_transform(products['default_category'])
tfidf_matrix.shape

(7636, 210)

### Pemodelan

##### Cosine Similarity

Sekarang, dihitung derajat kesamaan (similarity degree) antar rating dari kategori yang sama pada produk dengan teknik cosine similarity.

In [31]:
total_rating = products[['average_rating']].values
combined_features = np.hstack([total_rating, tfidf_matrix.toarray()])
cosine_sim_rating = cosine_similarity(combined_features)

Pada tahapan ini, dihitung cosine similarity dataframe tfidf_matrix yang diperoleh sebelumnya, lalu menyimpan nilai rating dalam variabel total_rating, dan menggabungkan dalam bentuk array dari hasil tfidf_matrix dengan nilai rating yang sudah disimpan

In [32]:
cosine_sim_df_2 = pd.DataFrame(cosine_sim_rating, index=products['product_name'], columns=products['product_name'])
print('Shape:', cosine_sim_df_2.shape)

cosine_sim_df_2.sample(20, axis=1).sample(20, axis=0)

Shape: (7636, 7636)


product_name                                        Lip & Eye Make Up Remover  \
product_name                                                                    
Kimchi Coolagen Cream Moisturizer                                    0.956198   
The Complexion Kit                                                   0.000000   
I AM CANCER BODY LOTION                                              0.953320   
Darling Lash out love + Curler                                       0.000000   
Deep Clean Hydrating Foaming Cleanser                                0.954735   
Carecoal - SUNDAE Shower Cream                                       0.954518   
Moongazing All-In-one Face Palette                                   0.957611   
Total Effects 7in1 Day Cream Gentle SPF 15                           0.952246   
Matte Coat + 2-in-1 Base & Top Coat • Duo Set                        0.000000   
Cranberry Juice Face Mask                                            0.951481   
Super Matcha Pore Tightening Toner                                   0.953519   
Mild Skin Balancing Vegan Toner                                      0.000000   
Spot-On Manicure Exceptionally Resilient                             0.957611   
Sold In Ceramide Lip Essence                                         0.000000   
Rice and Algae Blend Face Cleanser                                   0.957611   
English Pear Anti Acne and Scar Solution Body S...                   0.952335   
MAD FOR MAKEUP BT21 Spotless Second Skin Serum ...                   0.954363   
Hydrating set                                                        0.000000   
Iris restorative travel size shampoo + conditioner                   0.000000   
Glozee - Feminine Rose                                               0.000000   

product_name                                        Safi White Natural Brightening Cream Mangosteen  \
product_name                                                                                          
Kimchi Coolagen Cream Moisturizer                                                          0.999635   
The Complexion Kit                                                                         0.000000   
I AM CANCER BODY LOTION                                                                    0.963015   
Darling Lash out love + Curler                                                             0.000000   
Deep Clean Hydrating Foaming Cleanser                                                      0.964985   
Carecoal - SUNDAE Shower Cream                                                             0.951300   
Moongazing All-In-one Face Palette                                                         0.961852   
Total Effects 7in1 Day Cream Gentle SPF 15                                                 0.999968   
Matte Coat + 2-in-1 Base & Top Coat • Duo Set                                              0.000000   
Cranberry Juice Face Mask                                                                  0.948272   
Super Matcha Pore Tightening Toner                                                         0.950304   
Mild Skin Balancing Vegan Toner                                                            0.000000   
Spot-On Manicure Exceptionally Resilient                                                   0.954382   
Sold In Ceramide Lip Essence                                                               0.000000   
Rice and Algae Blend Face Cleanser                                                         0.966943   
English Pear Anti Acne and Scar Solution Body S...                                         0.962300   
MAD FOR MAKEUP BT21 Spotless Second Skin Serum ...                                         0.951145   
Hydrating set                                                                              0.000000   
Iris restorative travel size shampoo + conditioner                                         0.000000   
Glozee - Feminine Rose                                

Dengan cosine similarity, berhasil mengidentifikasi kesamaan antar produk lainnya. Shape (7636, 7636) merupakan ukuran matriks similarity dari data. Namun, dalam output, hanya menampilkan sebanyak 20 sampel data saja.


Contoh: angka 	0.961538 pada Papaya dan Clear Line N' Lash menunjukkan dua produk ini memiliki kesamaan kategori dan rating yang serupa

##### Mendapatkan Rekomendasi

Sistem rekomendasi akan memberikan produk yang memiliki similarity terhadap produk yang diinput oleh pengguna berdasarkan kesamaan kategori dan rating yang paling tinggi dari produk-produk rekomendasi, hasil similarity tiap produk sudah didapat dari perhitungan sebelumnya.

In [33]:
def products_recommendations_by_rating(nama_produk, similarity_data=cosine_sim_df_2, items=products[['product_name',
                                                                                                     'brand_name',
                                                                                                     'price_range',
                                                                                                     'average_rating']], k=10):
    index = similarity_data.loc[:,nama_produk].to_numpy().argpartition(
        range(-1, -k-1, -1))

    closest = similarity_data.columns[index[-1:-(k+2):-1]]
    closest = closest.drop(nama_produk, errors='ignore')
    recommendations = pd.DataFrame(closest, columns=['product_name']).merge(items, on='product_name')
    recommendations = recommendations.sort_values(by='average_rating', ascending=False)

    return recommendations.head(k)

Dibuat fungsi dengan nama products_recommendations, dengan nama_produk sebagai parameter pencarian, hasil kesamaan yang diambil dari cosine_sim_df, isi dari dataframe yang ingin ditampilkan dan k (jumlah rekomendasi yang diinginkan) sebanyak 10.

Lalu membuat index untuk mengambil urutan indek produk yang paling mirip, dengan similarity_data, dengan fungsi argpartition untuk mengurutkan indeks array berdasarkan skor kemiripan dari yang tertinggi ke terendah (berdasarkan parameter range(-1, -k-1, -1)).

Daftar produk disimpan dalam closest, dengan mengambil skor kemiripan tertinggi, lalu menghapus nama produk itu sendiri dari daftar rekomendasi. terakhir, dibuat variabel recommendation, untuk membuat dataframe dari data closest, column, dan items untuk digabung menjadi satu, diberikan tambahan untuk mengurutkan produk rekomendasi dari rating yang paling tinggi dan nilai k dikembalikan.

In [34]:
products['product_name'].unique()

array(['MULTI EYE COLOR PALETTE', 'VELVET LIP TINT', 'LIP COLOR', ...,
       'Repair Hair Mask', 'Hand Cream Olive Petitgrain', 'Color Mask'],
      dtype=object)

In [35]:
products[products.product_name.eq('Repair Hair Mask')]

brand_name      product_name  product_id price_range  \
7633  yves-rocher  Repair Hair Mask       36357  Rp 289.000   

                                                    url default_category  \
7633  https://www.sociolla.com/hair-mask/14833-repai...        Hair Mask   

      average_rating  total_reviews  total_recommended_count  \
7633            4.75              5                        5   

      total_in_wishlist brand_name_id  
7633                 58           273

Output rekomendasi diharapkan akan memberikan produk serupa Repair Hair Mask, dengan kategori Hair Mask yang mirip

In [36]:
products_recommendations_by_rating('Repair Hair Mask')

product_name             brand_name  \
8                      Bonheur Grasse Rose Treatment            moist-diane   
7       Roughness Eraser Shower Scrub flora's Secret                lavojoy   
2  Tresemme Keratin Deep Smoothening Hair Mask Pe...               tresemme   
0  MISSDAISY Perfume Hair Mask - Blackcurrant & V...             miss-daisy   
1                    Balancing & Soothing Scalp Pack            rated-green   
3  Hair Energy Fibertherapy Hair & Scalp Creambat...               makarizo   
4              Lador Tea Tree Scalp Clinic Hair Pack                  lador   
5  Texture Experience Creambath Vanilla Milk Inte...  makarizo-professional   
6                            Marula Repair Hair Mask                dancoly   
9                              Argan Repair Hairmask                dancoly   

              price_range  average_rating  
8              Rp 190.000        4.784091  
7               Rp 99.000        4.783898  
2               Rp 87.000        4.750000  
0              Rp 366.000        4.750000  
1              Rp 118.000        4.750000  
3  Rp 22.154 - Rp 124.100        4.745370  
4              Rp 210.000        4.733333  
5              Rp 288.600        4.732143  
6              Rp 329.000        4.718254  
9              Rp 289.000        4.708498

Output pun menampilkan 10 daftar produk rekomendasi yang memiliki kemiripan dengan Repair Hair Mask, dengan rentang harga, asal brandnya, dan nilai rating dari yang paling tinggi ke terendah

In [37]:
products_recommendations_by_rating('MULTI EYE COLOR PALETTE')

product_name  brand_name  \
6                                    Cute Eyes Maker       etude   
7                           Mirror Holic Liquid Eyes       etude   
8                            Look My Eyes Velvet NEW       etude   
9                         Play Color Eyes Mini Objet       etude   
4                           LIQUID PRIMER EYE SHADOW         3ce   
2                   lilybyred Skinny Mes Brow Pencil   lilybyred   
0                               Look At My Eyes Café       etude   
1                               Pro Eye Palette Mini        clio   
3                       MINI MULTI EYE COLOR PALETTE         3ce   
5  YOU Colorland Explorer Colorland Wander Nature...  you-beauty   

              price_range  average_rating  
6              Rp 205.000        5.000000  
7              Rp 188.000        5.000000  
8              Rp 109.000        5.000000  
9              Rp 342.000        5.000000  
4              Rp 264.000        4.975000  
2              Rp 120.000        4.933333  
0  Rp 79.000 - Rp 109.000        4.920000  
1              Rp 310.000        4.920000  
3              Rp 423.000        4.883333  
5              Rp 258.000        4.850000

## Model Development dengan Content Based Filtering berdasarkan Recommended Count

Teknik content based filtering akan merekomendasikan item yang mirip dengan item yang disukai pengguna di masa lalu. Pada tahap ini, ditemukan representasi fitur penting dari setiap kategori produk dengan tfidf vectorizer dan menghitung tingkat kesamaan berdasarkan total rekomendasi dengan cosine similarity.

### Data Preparation

##### TF-IDF Vectorizer

Teknik TF-IDF digunakan pada sistem rekomendasi untuk menemukan representasi fitur penting dari setiap kategori produk.

In [38]:
products

brand_name                  product_name  product_id  \
0             3ce       MULTI EYE COLOR PALETTE       97802   
1             3ce               VELVET LIP TINT       97810   
2             3ce                     LIP COLOR       97822   
3             3ce  MINI MULTI EYE COLOR PALETTE       97833   
4             3ce                    FACE BLUSH       97801   
...           ...                           ...         ...   
7631  yves-rocher    Ambre Noir Eau De Toilette        2158   
7632  yves-rocher  Soin Stimulating Conditioner       64174   
7633  yves-rocher              Repair Hair Mask       36357   
7634  yves-rocher   Hand Cream Olive Petitgrain       36404   
7635  yves-rocher                    Color Mask       64153   

                  price_range  \
0     Rp 555.000 - Rp 687.000   
1                  Rp 264.000   
2                  Rp 317.000   
3                  Rp 423.000   
4                  Rp 300.000   
...                       ...   
7631               Rp 659.000   
7632               Rp 199.000   
7633               Rp 289.000   
7634               Rp 109.000   
7635               Rp 289.000   

                                                    url   default_category  \
0     https://www.sociolla.com/eyeshadow/69460-phan-...          Eyeshadow   
1     https://www.sociolla.com/lip-cream/69468-son-k...          Lip Cream   
2     https://www.sociolla.com/lip-matte/69480-son-t...           Lipstick   
3     https://www.sociolla.com/eyeshadow/69491-phan-...          Eyeshadow   
4     https://www.sociolla.com/blush/69459-phan-ma-h...              Blush   
...                                                 ...                ...   
7631  https://sociolla.com/fragrance/8088-ambre-noir...    Eau De Toilette   
7632  https://vn.sociolla.com/d-u-x/17239-soin-stimu...        Conditioner   
7633  https://www.sociolla.com/hair-mask/14833-repai...          Hair Mask   
7634  https://www.sociolla.com/221-hand-and-foot-cre...  Hand & Foot Cream   
7635   https://vn.sociolla.com/kem-t-c/17224-color-mask          Hair Mask   

      average_rating  total_reviews  total_recommended_count  \
0           4.920000              5                        5   
1           4.576190             42                       42   
2           0.000000              0                        0   
3           4.883333              6                       12   
4           4.858824             13                       17   
...              ...            ...                      ...   
7631        4.476190              7                        6   
7632        4.638889              9                        7   
7633        4.750000              5                        5   
7634        4.968750              8                        8   
7635        4.500000              2                        2   

      total_in_wishlist brand_name_id  
0                   717           796  
1                   682           796  
2                   173           796  
3                   257           796  
4                   387           796  
...                 ...           ...  
7631                355           273  
7632                122           273  
7633                 58           273  
7634                117           273  
7635                 35           273  

[7636 rows x 11 columns]

In [39]:
from sklearn.feature_extraction.text import TfidfVectorizer

tf = TfidfVectorizer()
tf.fit(products['default_category'])
tf.get_feature_names_out()

array(['2in1', 'accessories', 'acne', 'after', 'ampoule', 'and',
       'applicators', 'aromatherapy', 'arts', 'baby', 'bags', 'balm',
       'bar', 'bath', 'bb', 'beauty', 'blotting', 'blush', 'body',
       'booster', 'bottles', 'breast', 'bronzer', 'brow', 'brush',
       'brushes', 'bug', 'bundles', 'butter', 'cake', 'cap', 'card',
       'care', 'case', 'cc', 'cellulite', 'clay', 'cleaner', 'cleanser',
       'cleansing', 'clipper', 'clippers', 'clothing', 'color', 'combs',
       'concealer', 'conditioner', 'contour', 'cotton', 'crayon', 'cream',
       'curler', 'curling', 'cushion', 'de', 'deodorant', 'diaper', 'dry',
       'dryers', 'eau', 'essence', 'exclusive', 'exfoliants',
       'exfoliator', 'eye', 'eyebrows', 'eyelash', 'eyelashes',
       'eyeliner', 'eyeshadow', 'fabric', 'face', 'facial', 'false',
       'family', 'feminine', 'files', 'foam', 'foot', 'for', 'foundation',
       'fragrance', 'gel', 'gift', 'gloss', 'glue', 'hair', 'hand',
       'head', 'highlighter'

In [40]:
tfidf_matrix = tf.fit_transform(products['default_category'])
tfidf_matrix.shape

(7636, 210)

### Pemodelan

##### Cosine Similarity

Sekarang, dihitung derajat kesamaan (similarity degree) antar total rekomendasi dari kategori yang sama pada produk dengan teknik cosine similarity.

In [41]:
total_recommended_count = products[['total_recommended_count']].values
combined_features = np.hstack([total_recommended_count, tfidf_matrix.toarray()])
cosine_sim_recom = cosine_similarity(combined_features)

Pada tahapan ini, dihitung cosine similarity dataframe tfidf_matrix yang diperoleh sebelumnya, lalu menyimpan total rekomendasi dalam variabel total_recommended_count, dan menggabungkan dalam bentuk array dari hasil tfidf_matrix dengan total rekomendasi yang sudah disimpan

In [42]:
cosine_sim_df_3 = pd.DataFrame(cosine_sim_recom, index=products['product_name'], columns=products['product_name'])
print('Shape:', cosine_sim_df_3.shape)

cosine_sim_df_3.sample(20, axis=1).sample(20, axis=0)

Shape: (7636, 7636)


product_name                                        YOUR SKIN BAE Toner Ceramide LC S-20 1% + Mugwort + Cica  \
product_name                                                                                                   
Renewing Body Lotion Collagen & Elastin                                                      0.999536          
Don't Forget Sunscreen                                                                       0.999991          
02 Air Mask Lemon Moisturizing & Brightening                                                 0.894423          
Green Tea Seed Eye and Face Ball                                                             0.894423          
Melt My Day! Cleansing Balm                                                                  0.999737          
NIVEA MEN Crème                                                                              0.999856          
AC Collection Ultimate Spot Cream                                                            0.999992          
Soothing & Calming Lotion Twin Set                                                           0.000000          
Vita Duo Cream Joan Day Joan Night                                                           0.999610          
[BUNDLE] Jacquelle Disney Princess Ariel editio...                                           0.995032          
Hydrium Green Tea Soothing Cream + Centella Soo...                                           0.948679          
Vaseline Sunblock SPF 30 Twinpack                                                            0.707103          
Double Pack Vitamin C                                                                        0.000000          
1606 Start The Day Beautiful 6Pc Set                                                         0.999911          
Ultra Cover Liquid Matte Foundation                                                          0.999993          
Twinpack Perfect UV Sunscreen Mild Milk 60ml                                                 0.000000          
Extra Glowing Kit                                                                            0.000000          
SKIN1004 Madagascar Centella Hyalu-Cica Brighte...                                           0.999996          
Feeling Good                                                                                 0.999401          
YOU Best Cushion Package                                                                     0.000000          

product_name                                        Gokujyun Ultimate Moisturizing Lotion  \
product_name                                                                                
Renewing Body Lotion Collagen & Elastin                                          0.999544   
Don't Forget Sunscreen                                                           0.999996   
02 Air Mask Lemon Moisturizing & Brightening                                     0.894427   
Green Tea Seed Eye and Face Ball                                                 0.894427   
Melt My Day! Cleansing Balm                                                      0.999742   
NIVEA MEN Crème                                                                  0.999867   
AC Collection Ultimate Spot Cream                                                0.999997   
Soothing & Calming Lotion Twin Set                                               0.000085   
Vita Duo Cream Joan Day Joan Night                                               0.999623   
[BUNDLE] Jacquelle Disney Princess Ariel editio...                               0.995037   
Hydrium Green Tea Soothing Cream + Centella Soo...                               0.948683   
Vaseline Sunblock SPF 30 Twinpack                                                0.707107   
Double Pack Vitamin C                                                            0.000000   
1606 Start The Day Beautiful 6Pc Set                                             0.999916   
Ultra Cover Liquid Matte Foundation            

Dengan cosine similarity, berhasil mengidentifikasi kesamaan antar produk lainnya. Shape (7636, 7636) merupakan ukuran matriks similarity dari data. Namun, dalam output, hanya menampilkan sebanyak 20 sampel data saja.


Contoh: angka 1.000000 pada Hand Butter Set (3pcs) dan Botanical Essentials - Bundle PATCHOULI Body Lotion and Hand Sanitizer menunjukkan dua produk ini memiliki kesamaan kategori dan total rekomendasi yang serupa

##### Mendapatkan Rekomendasi

Sistem rekomendasi akan memberikan produk yang memiliki similarity terhadap produk yang diinput oleh pengguna berdasarkan kesamaan kategori dan rating yang paling tinggi dari produk-produk rekomendasi, hasil similarity tiap produk sudah didapat dari perhitungan sebelumnya.

In [43]:
def products_recommendations_by_count_recom(nama_produk, similarity_data=cosine_sim_df_3, items=products[['product_name',
                                                                                                     'brand_name',
                                                                                                     'price_range',
                                                                                                     'total_recommended_count']], k=10):
    index = similarity_data.loc[:,nama_produk].to_numpy().argpartition(
        range(-1, -k-1, -1))

    closest = similarity_data.columns[index[-1:-(k+2):-1]]
    closest = closest.drop(nama_produk, errors='ignore')
    recommendations = pd.DataFrame(closest, columns=['product_name']).merge(items, on='product_name')
    recommendations = recommendations.sort_values(by='total_recommended_count', ascending=False)

    return recommendations.head(k)

Dibuat fungsi dengan nama products_recommendations, dengan nama_produk sebagai parameter pencarian, hasil kesamaan yang diambil dari cosine_sim_df, isi dari dataframe yang ingin ditampilkan dan k (jumlah rekomendasi yang diinginkan) sebanyak 10.

Lalu membuat index untuk mengambil urutan indek produk yang paling mirip, dengan similarity_data, dengan fungsi argpartition untuk mengurutkan indeks array berdasarkan skor kemiripan dari yang tertinggi ke terendah (berdasarkan parameter range(-1, -k-1, -1)).

Daftar produk disimpan dalam closest, dengan mengambil skor kemiripan tertinggi, lalu menghapus nama produk itu sendiri dari daftar rekomendasi. terakhir, dibuat variabel recommendation, untuk membuat dataframe dari data closest, column, dan items untuk digabung menjadi satu, diberikan tambahan untuk mengurutkan produk rekomendasi dari total rekomendasi yang paling banyak dan nilai k dikembalikan.

In [44]:
products['product_name'].unique()

array(['MULTI EYE COLOR PALETTE', 'VELVET LIP TINT', 'LIP COLOR', ...,
       'Repair Hair Mask', 'Hand Cream Olive Petitgrain', 'Color Mask'],
      dtype=object)

In [45]:
products[products.product_name.eq('Repair Hair Mask')]

brand_name      product_name  product_id price_range  \
7633  yves-rocher  Repair Hair Mask       36357  Rp 289.000   

                                                    url default_category  \
7633  https://www.sociolla.com/hair-mask/14833-repai...        Hair Mask   

      average_rating  total_reviews  total_recommended_count  \
7633            4.75              5                        5   

      total_in_wishlist brand_name_id  
7633                 58           273

Output rekomendasi diharapkan akan memberikan produk serupa Repair Hair Mask, dengan kategori Hair Mask yang mirip

In [46]:
products_recommendations_by_count_recom('Repair Hair Mask')

product_name             brand_name  \
7  Texture Experience Creambath Mint Sorbet Purif...  makarizo-professional   
8                             Wind Down Scalp Masque            runa-beauty   
6               Nourishing & Moisturizing Scalp Pack            rated-green   
4  Texture Experience Creambath Green Tea Butter ...  makarizo-professional   
5  Tresemme Keratin Deep Smoothening Hair Mask Pe...               tresemme   
2                              Miracle You Treatment            moist-diane   
3     Herb Mask Deep Conditioning & Repairing Damage     ree-derma-wellness   
1                      Extra Moist & Shine Hair Mask            moist-diane   
0                       Honey Dew Repair Mask Dusset  makarizo-professional   
9  Hair Energy Fibertherapy Hair & Scalp Creambat...               makarizo   

  price_range  total_recommended_count  
7  Rp 288.600                       13  
8  Rp 199.000                       13  
6  Rp 118.000                       10  
4  Rp 288.600                        9  
5   Rp 87.000                        9  
2  Rp 149.000                        7  
3  Rp 118.000                        7  
1  Rp 119.000                        6  
0  Rp 409.600                        5  
9  Rp 124.100                        3

Output pun menampilkan 10 daftar produk rekomendasi yang memiliki kemiripan dengan Repair Hair Mask, dengan rentang harga, asal brandnya, dan total rekomendasi dari yang paling banyak ke yang sedikit

In [47]:
products_recommendations_by_count_recom('VELVET LIP TINT')

product_name            brand_name  \
9                        Misty Matte Lip Cream             dazzle-me   
8                       CREAMY LIP & CHEEK DUO             focallure   
6                 Colorfit Fresh Matte Lip Ink                wardah   
5    Exclusive Matte Lip Cream X Ayang Cempaka                wardah   
4  YOU NEW FORMULA Rouge Power Matte Lip Cream            you-beauty   
3                       Silky Velvet Lip Cream             pinkflash   
1                             Matte Lip Colour  sada-by-cathy-sharon   
0                       Velvet Matte Lip Cream             dazzle-me   
2                        Staymax Matte Lip Ink             focallure   
7                       Lasting Matte Lipcream             pinkflash   

  price_range  total_recommended_count  
9   Rp 33.900                       76  
8   Rp 38.000                       73  
6   Rp 64.000                       68  
5   Rp 66.500                       67  
4   Rp 93.000                       51  
3   Rp 70.000                       45  
1  Rp 180.000                       44  
0   Rp 29.900                       42  
2   Rp 89.000                       40  
7   Rp 42.000                       30